data_structures/skip_list.cpp

```diff
@@ -0,0 +1,197 @@
+/**
+ * A skip list is a data structure that is used for storing a sorted list of items with a
+ *  help of hierarchy of linked lists that connect increasingly sparse subsequences of the items
+ * 
+ *  References used: GeeksForGeeks skip list code, https://iq.opengenus.org/skip-list/ Pseudo Code.
+*/
+
+#include <iostream>
+#include <vector>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <time.h>   
+#include <cstdlib>
+
+// using namespace std; 
+using std::vector;
+using std::endl;
+
+#define MAXLVL 2
+#define P 0.5
+  
+
+/**
+ *  Node structure [Key][Node*, Node*...]
+*/
+class Node { 
+public: 
+    int key; 
+    void* value;
+    /*Forward Array*/
+    vector<Node*> forward;
+    Node(int key, int level, void* value); 
+};
+
+  
+Node::Node(int key, int level, void* value) { 
+    this->key = key; 
+  
+    /*Initialization of forward vector*/
+    for (int i = 0; i < sizeof(Node*)*(level+1); i++){
+        forward.push_back(NULL);
+    }
+} 
+  
+// Class for Skip list 
+class SkipList { 
+    int level; 
+    Node *header; 
+public: 
+    SkipList(); 
+    int randomLevel(); 
+    void insertElement(int, void*); 
+    void deleteElement(int); 
+    void* searchElement(int); 
+    void displayList(); 
+};
+  
+SkipList::SkipList() { 
+    level = 0; 
+    /* Header initialization*/
+    header = new Node(-1, MAXLVL, NULL); 
+} 
+
+
+
+  
+/**
+ * Returns random level for skip list;
+*/
+int SkipList::randomLevel() { 
+    int lvl = 0; 
+    while(static_cast<float>(std::rand())/RAND_MAX < P && lvl < MAXLVL) lvl++;  
+    return lvl; 
+} 
+  
+  
+// Insert given key in skip list 
+void SkipList::insertElement(int key, void* value) { 
+    std::cout << "Inserting" << key << "...";
+    Node *x = header; 
+    Node *update[MAXLVL+1]; 
```

Comment:
here, `update` is a *pointer to* an *array of* `Node` data type. Is this the intent?

Reply:
yes

Reply:
in that case line# 103 doesn't make sense and my comment there stands.

Reply:
I misinterpret your first remark. update is an array of Node pointers.

Reply:
as mentioned in the comment above, this line crates `update` as a pointer of pointer. That is not what you wanted to do .

Reply:
after reviewing, it seems to be correct. My apologies. 
update is indeed supposed to be a an array of pointers.